plugins/ccgx/fu-ccgx-native-hid-device.c

```diff
@@ -0,0 +1,505 @@
+	}
+	fu_progress_step_done(progress);
+
+	// TODO: Doing this and reset by themselves (with magic unlock) doesn't
+	// switch to the alternative image. Seems we always need to flash in
+	// order to switch.
+	g_debug("before bootswitch");
+	// TODO: fw_img_no should just be
+	// fw_img_no = fu_ccgx_firmware_get_fw_mode(FW_CCGX_FIRMWARE(firmware))
+	if (!fu_ccgx_native_hid_command(self, CCGX_HID_CMD_SET_BOOT, fw_img_no, error)) {
+		g_prefix_error(error, "bootswitch command error: ");
+		return FALSE;
+	}
+	g_debug("After bootswitch");
+	fu_progress_step_done(progress);
+
+	g_debug("before reset");
+	if (!fu_ccgx_native_hid_command(self,
+					CCGX_HID_CMD_JUMP,
+					FU_CCGX_PD_RESP_DEVICE_RESET_CMD_SIG,
+					error)) {
+		g_prefix_error(error, "reset command error: ");
+		return FALSE;
+	}
+	g_debug("After reset");
+
+	return TRUE;
+}
+
+static FuFirmware *
+fu_ccgx_native_hid_device_prepare_firmware(FuDevice *device,
+					   GBytes *fw,
+					   FwupdInstallFlags flags,
+					   GError **error)
+{
+	FuCcgxNativeHidDevice *self = FU_CCGX_NATIVE_HID_DEVICE(device);
+	FuCcgxFwMode fw_mode;
+	guint16 fw_silicon_id;
+	g_autoptr(FuFirmware) firmware = fu_ccgx_firmware_new();
+
+	/* parse all images */
+	if (!fu_firmware_parse(firmware, fw, flags, error))
+		return NULL;
+
+	/* check the silicon ID */
+	fw_silicon_id = fu_ccgx_firmware_get_silicon_id(FU_CCGX_FIRMWARE(firmware));
+	if (fw_silicon_id != self->silicon_id) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "silicon id mismatch, expected 0x%x, got 0x%x",
+			    self->silicon_id,
+			    fw_silicon_id);
+		return NULL;
+	}
+	// TODO: Check if we can get this from the device at run-time
+	// if ((flags & FWUPD_INSTALL_FLAG_IGNORE_VID_PID) == 0) {
+	//	fw_app_type = fu_ccgx_firmware_get_app_type(FU_CCGX_FIRMWARE(firmware));
+	//	if (fw_app_type != self->fw_app_type) {
+	//		g_set_error(error,
+	//			    FWUPD_ERROR,
+	//			    FWUPD_ERROR_NOT_SUPPORTED,
+	//			    "app type mismatch, expected 0x%x, got 0x%x",
+	//			    self->fw_app_type,
+	//			    fw_app_type);
+	//		return NULL;
+	//	}
+	//}
+	fw_mode = fu_ccgx_firmware_get_fw_mode(FU_CCGX_FIRMWARE(firmware));
+	if (fw_mode != fu_ccgx_fw_mode_get_alternate(self->fw_mode)) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "FuCcgxFwMode mismatch, expected %s, got %s",
+			    fu_ccgx_fw_mode_to_string(fu_ccgx_fw_mode_get_alternate(self->fw_mode)),
+			    fu_ccgx_fw_mode_to_string(fw_mode));
+		return NULL;
+	}
+	return g_steal_pointer(&firmware);
+}
+
+static gboolean
+fu_ccgx_native_hid_device_write_firmware(FuDevice *device,
+					 FuFirmware *firmware,
+					 FuProgress *progress,
+					 FwupdInstallFlags flags,
+					 GError **error)
+{
+	FuCcgxNativeHidDevice *self = FU_CCGX_NATIVE_HID_DEVICE(device);
+
+	g_debug("Operating Mode:  0x%02x", self->fw_mode);
+	switch (self->fw_mode) {
+	case FU_CCGX_FW_MODE_BOOT:
+	case FU_CCGX_FW_MODE_FW2:
```

Comment:
can you just do `if (self->fw_mode == FU_CCGX_FW_MODE_BOOT || self->fw_mode == FU_CCGX_FW_MODE_FW2)` instead please and then move out the common code.